profiler/src/ProfilerEngine/Datadog.Profiler.Native/ExceptionsProvider.cpp

```diff
@@ -0,0 +1,252 @@
+        return false;
+    }
+
+    if (assemblyName != "System.Private.CoreLib" && assemblyName != "mscorlib")
+    {
+        return false;
+    }
+
+    _mscorlibModuleId = moduleId;
+
+    INVOKE(_pCorProfilerInfo->GetStringLayout2(&_stringLengthOffset, &_stringBufferOffset))
+
+    return true;
+}
+
+bool ExceptionsProvider::OnExceptionThrown(ObjectID thrownObjectId)
+{
+    if (_mscorlibModuleId == 0)
+    {
+        if (!_loggedMscorlibError)
+        {
+            _loggedMscorlibError = true;
+            Log::Warn("An exception has been thrown but mscorlib is not loaded");
+        }
+        return false;
+    }
+
+    if (_exceptionClassId == 0)
+    {
+        if (!LoadExceptionMetadata())
+        {
+            return false;
+        }
+    }
+
+    ClassID classId;
+
+    INVOKE(_pCorProfilerInfo->GetClassFromObject(thrownObjectId, &classId))
+
+    std::string name;
+
+    if (!GetExceptionType(classId, name))
+    {
+        return false;
+    }
+
+    const auto messageAddress = *reinterpret_cast<UINT_PTR*>(thrownObjectId + _messageFieldOffset.ulOffset);
+
+    std::string message;
+
+    if (messageAddress == 0)
+    {
+        message = std::string();
+    }
+    else
+    {
+        const auto stringLength = *reinterpret_cast<ULONG*>(messageAddress + _stringLengthOffset);
+
+        if (stringLength == 0)
+        {
+            message = std::string();
+        }
+        else
+        {
+            message = shared::ToString(reinterpret_cast<WCHAR*>(messageAddress + _stringBufferOffset), stringLength);
+        }
+    }
+
+    ManagedThreadInfo* threadInfo;
+
+    INVOKE(_pManagedThreadList->TryGetCurrentThreadInfo(&threadInfo))
+
+    uint32_t hrCollectStack = E_FAIL;
+    const auto pStackFramesCollector = OsSpecificApi::CreateNewStackFramesCollectorInstance(_pCorProfilerInfo);
+
+    pStackFramesCollector->PrepareForNextCollection();
+    const auto result = pStackFramesCollector->CollectStackSample(threadInfo, &hrCollectStack);
+
+    if (result->GetFramesCount() == 0)
+    {
+        Log::Warn("Failed to walk stack for thrown exception: ", HResultConverter::ToStringWithCode(hrCollectStack));
+        return false;
+    }
+
+    result->DetermineAppDomain(threadInfo->GetClrThreadId(), _pCorProfilerInfo);
+
+    RawExceptionSample rawSample;
+
+    rawSample.Timestamp = result->GetUnixTimeUtc();
+    rawSample.LocalRootSpanId = result->GetLocalRootSpanId();
+    rawSample.SpanId = result->GetSpanId();
+    rawSample.AppDomainId = result->GetAppDomainId();
+    result->CopyInstructionPointers(rawSample.Stack);
+    rawSample.ThreadInfo = threadInfo;
+    threadInfo->AddRef();
+    rawSample.ExceptionMessage = std::move(message);
+    rawSample.ExceptionType = std::move(name);
+    Add(std::move(rawSample));
+
+    return true;
+}
+
+bool ExceptionsProvider::GetExceptionType(ClassID classId, std::string& exceptionType)
+{
+    {
+        std::lock_guard lock(_exceptionTypesLock);
```

Comment:
Any risk of reentrancy leading to dead-lock?

Reply:
I made the scope as small as possible, it's only guarding the access to the map. No risk of reentrancy.